File: chat-gui/src/main/java/ru/usedesk/chat_gui/external/UsedeskChatFragment.java
package ru.usedesk.chat_gui.external;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModelProviders;

import com.google.android.material.bottomsheet.BottomSheetDialog;

import java.util.List;
import java.util.Set;

import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;
import ru.usedesk.chat_gui.R;
import ru.usedesk.chat_gui.internal.chat.AttachedFilesAdapter;
import ru.usedesk.chat_gui.internal.chat.ChatViewModel;
import ru.usedesk.chat_gui.internal.chat.ChatViewModelFactory;
import ru.usedesk.chat_gui.internal.chat.FilePicker;
import ru.usedesk.chat_gui.internal.chat.MessagePanelAdapter;
import ru.usedesk.chat_gui.internal.chat.MessagesAdapter;
import ru.usedesk.chat_gui.internal.chat.OfflineFormExpectedAdapter;
import ru.usedesk.chat_gui.internal.chat.OfflineFormSentAdapter;
import ru.usedesk.chat_sdk.external.UsedeskChatSdk;
import ru.usedesk.chat_sdk.external.entity.UsedeskFileInfo;
import ru.usedesk.chat_sdk.external.entity.UsedeskMessage;
import ru.usedesk.common_gui.external.UsedeskViewCustomizer;
import ru.usedesk.common_sdk.external.entity.exceptions.UsedeskException;

@RuntimePermissions
public class UsedeskChatFragment extends Fragment {
    private MessagePanelAdapter messagePanelAdapter;
    private MessagesAdapter messagesAdapter;
    private OfflineFormExpectedAdapter offlineFormExpectedAdapter;
    private OfflineFormSentAdapter offlineFormSentAdapter;
    private AttachedFilesAdapter attachedFilesAdapter;

    private TextView tvLoading;
    private ViewGroup ltContent;

    private FilePicker filePicker;

    private ChatViewModel viewModel;

    public static UsedeskChatFragment newInstance() {
        return new UsedeskChatFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        filePicker = new FilePicker();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = UsedeskViewCustomizer.getInstance()
                .createView(inflater, R.layout.usedesk_fragment_chat, container, false, R.style.Usedesk_Theme_Chat);

        viewModel = ViewModelProviders.of(this, new ChatViewModelFactory(container.getContext()))
                .get(ChatViewModel.class);

        initUi(view);
        renderData();
        observeData(getViewLifecycleOwner());

        return view;
    }

    private void initUi(@NonNull View view) {
        tvLoading = view.findViewById(R.id.tv_loading);
        ltContent = view.findViewById(R.id.lt_content);

        attachedFilesAdapter = new AttachedFilesAdapter(viewModel, view.findViewById(R.id.rv_attached_files));
        messagePanelAdapter = new MessagePanelAdapter(view, viewModel, v -> openAttachmentDialog(),
                getViewLifecycleOwner());
        offlineFormExpectedAdapter = new OfflineFormExpectedAdapter(view, viewModel, getViewLifecycleOwner());
        offlineFormSentAdapter = new OfflineFormSentAdapter(view, viewModel, getViewLifecycleOwner());
        messagesAdapter = new MessagesAdapter(view, viewModel,
                viewModel.getMessagesLiveData().getValue(),
                viewModel.getFeedbacksLiveData().getValue());
    }

    private void renderData() {
        onMessages(viewModel.getMessagesLiveData().getValue());
        onFileInfoList(viewModel.getFileInfoListLiveData().getValue());
    }

    private void observeData(@NonNull LifecycleOwner lifecycleOwner) {
        viewModel.getMessagesLiveData()
                .observe(lifecycleOwner, this::onMessages);

        viewModel.getFileInfoListLiveData()
                .observe(lifecycleOwner, this::onFileInfoList);

        viewModel.getFeedbacksLiveData()
                .observe(lifecycleOwner, this::onFeedbacks);

        viewModel.getExceptionLiveData()
                .observe(lifecycleOwner, this::onException);
    }

    private void openAttachmentDialog() {
        final BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(getContext());
        View bottomSheetView = UsedeskViewCustomizer.getInstance()
                .createView(getActivity().getLayoutInflater(),
                        R.layout.usedesk_dialog_attachment, null, false, R.style.Usedesk_Theme_Chat);

        bottomSheetView.findViewById(R.id.pick_photo_button)
                .setOnClickListener(view -> {
                    bottomSheetDialog.dismiss();
                    onPickPhotoClicked();
                });

        bottomSheetView.findViewById(R.id.take_photo_button)
                .setOnClickListener(view -> {
                    bottomSheetDialog.dismiss();
                    onTakePhotoClicked();
                });

        bottomSheetView.findViewById(R.id.pick_document_button)
                .setOnClickListener(view -> {
                    bottomSheetDialog.dismiss();
                    onPickDocumentClicked();
                });

        bottomSheetDialog.setContentView(bottomSheetView);
        bottomSheetDialog.show();
    }

    private void onException(UsedeskException exception) {
        if (exception != null) {
            String message = exception.getMessage();
            if (message == null) {
                message = exception.toString();
            }
            Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
        }
    }

    private void onFeedbacks(@Nullable Set<Integer> feedbacks) {
        if (feedbacks != null) {
            messagesAdapter.updateFeedbacks(feedbacks);
        }
    }

    private void onMessages(@Nullable List<UsedeskMessage> usedeskMessages) {
        boolean isMessages = usedeskMessages != null;
        tvLoading.setVisibility(isMessages
                ? View.GONE
                : View.VISIBLE);
        ltContent.setVisibility(isMessages
                ? View.VISIBLE
                : View.GONE);

        if (usedeskMessages != null) {
            messagesAdapter.updateMessages(usedeskMessages);
        }
    }

    private void onFileInfoList(@Nullable List<UsedeskFileInfo> usedeskFileInfoList) {
        if (usedeskFileInfoList != null) {
            attachedFilesAdapter.update(usedeskFileInfoList);
        }
    }

    private void onPickPhotoClicked() {
        UsedeskChatFragmentPermissionsDispatcher.pickPhotoWithPermissionCheck(this);
    }

    private void onTakePhotoClicked() {
        UsedeskChatFragmentPermissionsDispatcher.takePhotoWithPermissionCheck(this);
    }

    private void onPickDocumentClicked() {
        UsedeskChatFragmentPermissionsDispatcher.pickDocumentWithPermissionCheck(this);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        UsedeskChatFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    public void onStart() {
        super.onStart();

        UsedeskChatSdk.stopService(getContext());
    }

    @Override
    public void onStop() {
        super.onStop();

        UsedeskChatSdk.startService(getContext());
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            List<UsedeskFileInfo> attachedFileInfoList = filePicker.onResult(getContext(),
                    requestCode, data);
            if (attachedFileInfoList != null) {
                viewModel.setAttachedFileInfoList(attachedFileInfoList);
            }
        }
    }

    @NeedsPermission({Manifest.permission.WRITE_EXTERNAL_STORAGE})
    void pickPhoto() {
        filePicker.pickImage(this);
    }

    @NeedsPermission({Manifest.permission.CAMERA})
    void takePhoto() {
        filePicker.takePhoto(this);
    }

    @NeedsPermission({Manifest.permission.WRITE_EXTERNAL_STORAGE})
    void pickDocument() {
        filePicker.pickDocument(this);
    }
}

File: chat-gui/src/main/java/ru/usedesk/chat_gui/external/UsedeskChatFragment.kt
package ru.usedesk.chat_gui.external

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.viewModels
import ru.usedesk.chat_gui.R
import ru.usedesk.chat_gui.databinding.UsedeskScreenChatBinding
import ru.usedesk.chat_gui.external.attachpanel.UsedeskDialogAttachmentPanel
import ru.usedesk.chat_gui.internal.chat.*
import ru.usedesk.chat_sdk.external.UsedeskChatSdk
import ru.usedesk.chat_sdk.external.entity.UsedeskFileInfo
import ru.usedesk.common_gui.external.UsedeskToolbar
import ru.usedesk.common_gui.internal.UsedeskFragment
import ru.usedesk.common_gui.internal.inflateBinding
import ru.usedesk.common_gui.internal.showInstead

class UsedeskChatFragment : UsedeskFragment(R.style.Usedesk_Theme_Chat) {

    private val viewModel: ChatViewModel by viewModels()

    private lateinit var binding: UsedeskScreenChatBinding
    private lateinit var attachmentDialog: UsedeskDialogAttachmentPanel

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        doInit {
            val agentName: String? = argsGetString(AGENT_NAME_KEY)

            binding = inflateBinding(inflater,
                    container,
                    R.layout.usedesk_screen_chat,
                    defaultStyleId)

            val title = getStringFromStyle(R.attr.usedesk_screen_chat_title)

            init(agentName)
            UsedeskToolbar(requireActivity() as AppCompatActivity, binding.toolbar).apply {
                setTitle(title)
                setBackButton {
                    requireActivity().onBackPressed()
                }
            }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        attachmentDialog = UsedeskDialogAttachmentPanel(this)
    }

    private fun init(agentName: String?) {
        UsedeskChatSdk.init(requireContext(), viewModel.actionListenerRx)

        viewModel.init()

        MessagePanelAdapter(binding.messagePanel, viewModel, viewLifecycleOwner) {
            openAttachmentDialog()
        }

        OfflineFormExpectedAdapter(binding.root, viewModel, viewLifecycleOwner)
        OfflineFormSentAdapter(binding.root, viewModel, viewLifecycleOwner)

        ItemsAdapter(viewModel,
                binding.rvMessages,
                agentName,
                viewLifecycleOwner, { file ->
            getParentListener<IUsedeskOnFileClickListener>()?.onFileClick(file)
        }, { html ->
            getParentListener<IUsedeskOnHtmlClickListener>()?.onHtmlClick(html)
        }, { url ->
            getParentListener<IUsedeskOnUrlClickListener>()?.onUrlClick(url)
                    ?: onUrlClick(url)
        })
        viewModel.feedbacksLiveData.observe(viewLifecycleOwner) {
            onFeedbacks(it)
        }
        viewModel.exceptionLiveData.observe(viewLifecycleOwner) {
            onException(it)
        }
        viewModel.chatItemsLiveData.observe(viewLifecycleOwner) {
            showInstead(binding.ltContent, binding.tvLoading, it != null)
        }
    }

    private fun onUrlClick(url: String) {
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    private inline fun <reified T> getParentListener(): T? {
        return when {
            activity is T -> {
                activity as T
            }
            parentFragment is T -> {
                parentFragment as T
            }
            else -> {
                null
            }
        }
    }

    private fun openAttachmentDialog() {
        getParentListener<IUsedeskOnAttachmentClickListener>()?.onAttachmentClick()
                ?: attachmentDialog.show()
    }

    private fun onException(exception: Exception) {
        exception.printStackTrace()
    }

    private fun onFeedbacks(feedbacks: Set<Int>?) {
        /*if (feedbacks != null) {
            messagesAdapter.updateFeedbacks(feedbacks);
        }*/
    }

    override fun onStart() {
        super.onStart()
        UsedeskChatSdk.stopService(requireContext())
    }

    override fun onStop() {
        super.onStop()
        UsedeskChatSdk.startService(requireContext())
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        attachmentDialog.onActivityResult(requestCode, resultCode, data)
    }

    fun setAttachedFiles(attachedFiles: List<UsedeskFileInfo>) {
        viewModel.setAttachedFiles(attachedFiles)
    }

    companion object {
        private const val AGENT_NAME_KEY = "agentNameKey"

        @JvmOverloads
        @JvmStatic
        fun newInstance(agentName: String? = null): UsedeskChatFragment {
            return UsedeskChatFragment().apply {
                arguments = Bundle().apply {
                    if (agentName != null) {
                        putString(AGENT_NAME_KEY, agentName)
                    }
                }
            }
        }
    }
}